File: bestia-models/src/test/java/net/bestia/model/map/MapChunkTest.java
package net.bestia.model.map;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.junit.Assert;
import org.junit.Test;

import net.bestia.model.geometry.Point;
import net.bestia.model.geometry.Rect;

public class MapChunkTest {

	private int[] groundLayer = new int[MapChunk.MAP_CHUNK_SIZE_AREA];
	private int[] groundLayerBig = new int[MapChunk.MAP_CHUNK_SIZE_AREA + 10];

	public MapChunkTest() {

		for (int i = 0; i < groundLayer.length; i++) {
			groundLayer[i] = 10;
		}

		for (int i = 0; i < groundLayerBig.length; i++) {
			groundLayerBig[i] = 10;
		}
	}

	@Test
	public void ctor1_ok() {
		new MapChunk(new Point(5, 4), groundLayer);
	}

	@Test(expected = IllegalArgumentException.class)
	public void ctor1_layerTooBig_throws() {
		new MapChunk(new Point(5, 4), groundLayerBig);
	}

	@Test(expected = IllegalArgumentException.class)
	public void ctor1_negPoint_throws() {
		new MapChunk(new Point(-5, 4), groundLayer);
	}

	@Test
	public void getGid_outOfRange_minusOne() {
		MapChunk mc = new MapChunk(new Point(5, 4), groundLayer);
		Assert.assertEquals(-1, mc.getGid(new Point(-10, 3)));
	}

	@Test
	public void getGid_outOfLayers_minusOne() {
		MapChunk mc = new MapChunk(new Point(5, 4), groundLayer);
		Assert.assertEquals(-1, mc.getGid(10, new Point(-10, 3)));
		Assert.assertEquals(-1, mc.getGid(-3, new Point(-10, 3)));
	}

	@Test
	public void getGid_ok() {
		List<Map<Point, Integer>> layers = new ArrayList<>();
		Map<Point, Integer> layer1 = new HashMap<>();
		layers.add(layer1);
		layer1.put(new Point(2, 3), 10);
		layer1.put(new Point(6, 2), 11);
		
		MapChunk mc = new MapChunk(new Point(5, 4), groundLayer, layers);
		
		Assert.assertEquals(10, mc.getGid(new Point(1, 1)));
		Assert.assertEquals(-1, mc.getGid(1, new Point(1, 1)));
		Assert.assertEquals(10, mc.getGid(1, new Point(2, 3)));
		Assert.assertEquals(11, mc.getGid(1, new Point(6, 2)));
		// Ground
		Assert.assertEquals(10, mc.getGid(0, new Point(1, 1)));
	}

	@Test(expected = IllegalArgumentException.class)
	public void getChunkCords_negative_throws() {
		MapChunk.Companion.getChunkCords(new Point(-1, 5));
	}

	@Test
	public void getChunkCords_ok() {
		Point p = MapChunk.Companion.getChunkCords(new Point(105, 5));
		Assert.assertEquals(new Point(5, 5), p);
	}

	@Test(expected = IllegalArgumentException.class)
	public void getWorldCords_negative_throws() {
		MapChunk.Companion.getWorldCords(new Point(-1, 0));
	}

	@Test
	public void getWorldCords_ok() {
		Point p = MapChunk.Companion.getWorldCords(new Point(5, 5));
		Assert.assertEquals(new Point(50, 50), p);
	}

	@Test(expected = IllegalArgumentException.class)
	public void getWorldRect_negative_throws() {
		MapChunk.Companion.getWorldRect(new Point(-1, 5));
	}

	@Test
	public void getWorldRect_ok() {
		Rect r = MapChunk.Companion.getWorldRect(new Point(1, 5));
		Assert.assertEquals(new Rect(10, 50, 10, 10), r);
	}
}

File: bestia-models/src/test/kotlin/net/bestia/model/map/MapChunkTest.kt
package net.bestia.model.map

import java.util.ArrayList
import java.util.HashMap

import org.junit.Assert
import org.junit.Test

import net.bestia.model.geometry.Point
import net.bestia.model.geometry.Rect

class MapChunkTest {

  private val groundLayer = IntArray(MapChunk.MAP_CHUNK_SIZE_AREA)
  private val groundLayerBig = IntArray(MapChunk.MAP_CHUNK_SIZE_AREA + 10)

  init {

    for (i in groundLayer.indices) {
      groundLayer[i] = 10
    }

    for (i in groundLayerBig.indices) {
      groundLayerBig[i] = 10
    }
  }

  @Test
  fun ctor1_ok() {
    MapChunk(Point(5, 4), groundLayer)
  }

  @Test(expected = IllegalArgumentException::class)
  fun ctor1_layerTooBig_throws() {
    MapChunk(Point(5, 4), groundLayerBig)
  }

  @Test(expected = IllegalArgumentException::class)
  fun ctor1_negPoint_throws() {
    MapChunk(Point(-5, 4), groundLayer)
  }

  @Test
  fun getGid_outOfRange_minusOne() {
    val mc = MapChunk(Point(5, 4), groundLayer)
    Assert.assertEquals(-1, mc.getGid(Point(-10, 3)).toLong())
  }

  @Test
  fun getGid_outOfLayers_minusOne() {
    val mc = MapChunk(Point(5, 4), groundLayer)
    Assert.assertEquals(-1, mc.getGid(10, Point(-10, 3)).toLong())
    Assert.assertEquals(-1, mc.getGid(-3, Point(-10, 3)).toLong())
  }

  @Test
  fun getGid_ok() {
    val layers = ArrayList<Map<Point, Int>>()
    val layer1 = HashMap<Point, Int>()
    layers.add(layer1)
    layer1[Point(2, 3)] = 10
    layer1[Point(6, 2)] = 11

    val mc = MapChunk(Point(5, 4), groundLayer, layers)

    Assert.assertEquals(10, mc.getGid(Point(1, 1)).toLong())
    Assert.assertEquals(-1, mc.getGid(1, Point(1, 1)).toLong())
    Assert.assertEquals(10, mc.getGid(1, Point(2, 3)).toLong())
    Assert.assertEquals(11, mc.getGid(1, Point(6, 2)).toLong())
    // Ground
    Assert.assertEquals(10, mc.getGid(0, Point(1, 1)).toLong())
  }

  @Test(expected = IllegalArgumentException::class)
  fun getChunkCords_negative_throws() {
    MapChunk.getChunkCords(Point(-1, 5))
  }

  @Test
  fun getChunkCords_ok() {
    val p = MapChunk.getChunkCords(Point(105, 5))
    Assert.assertEquals(Point(5, 5), p)
  }

  @Test(expected = IllegalArgumentException::class)
  fun getWorldCords_negative_throws() {
    MapChunk.getWorldCords(Point(-1, 0))
  }

  @Test
  fun getWorldCords_ok() {
    val p = MapChunk.getWorldCords(Point(5, 5))
    Assert.assertEquals(Point(50, 50), p)
  }

  @Test(expected = IllegalArgumentException::class)
  fun getWorldRect_negative_throws() {
    MapChunk.getWorldRect(Point(-1, 5))
  }

  @Test
  fun getWorldRect_ok() {
    val r = MapChunk.getWorldRect(Point(1, 5))
    Assert.assertEquals(Rect(10, 50, 10, 10), r)
  }
}
